drivers/sensor/pixart/paj7620/paj7620.c

```diff
@@ -0,0 +1,295 @@
+
+	/* Part ID is stored in bank 0 */
+	ret = paj7620_select_register_bank(dev, PAJ7620_MEMBANK_0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, PAJ7620_REG_PART_ID_LSB, &hw_id[0]);
+	if (ret < 0) {
+		LOG_ERR("Failed to read hardware ID");
+		return ret;
+	}
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, PAJ7620_REG_PART_ID_MSB, &hw_id[1]);
+	if (ret < 0) {
+		LOG_ERR("Failed to read hardware ID");
+		return ret;
+	}
+
+	*result = sys_get_le16(hw_id);
+	LOG_DBG("Obtained hardware ID 0x%04x", *result);
+
+	return 0;
+}
+
+static int paj7620_write_initial_reg_settings(const struct device *dev)
+{
+	int ret;
+	uint8_t reg_addr;
+	uint8_t value;
+	const struct paj7620_config *config = dev->config;
+
+	/**
+	 * Initializes registers with default values according to section 8.1
+	 * from Datasheet v1.5:
+	 * https://files.seeedstudio.com/wiki/Grove_Gesture_V_1.0/res/PAJ7620U2_DS_v1.5_05012022_Confidential.pdf
+	 */
+
+	for (int i = 0; i < ARRAY_SIZE(initial_register_array); i++) {
+		reg_addr = initial_register_array[i][0];
+		value = initial_register_array[i][1];
+
+		ret = i2c_reg_write_byte_dt(&config->i2c, reg_addr, value);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+static int paj7620_set_sampling_rate(const struct device *dev, const struct sensor_value *val)
+{
+	int ret;
+	int fps;
+	const struct paj7620_config *config = dev->config;
+	int64_t uval = val->val1 * 1000000 + val->val2;
+
+	if (uval <= 120000000) {
+		fps = PAJ7620_NORMAL_SPEED;
+	} else if (uval <= 240000000) {
+		fps = PAJ7620_GAME_SPEED;
+	} else {
+		LOG_ERR("Unsupported sample rate");
+		return -ENOTSUP;
+	}
+
+	ret = paj7620_select_register_bank(dev, PAJ7620_MEMBANK_1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, PAJ7620_REG_R_IDLE_TIME_LSB, fps);
+	if (ret < 0) {
+		LOG_ERR("Failed to set sample rate");
+		return ret;
+	}
+
+	ret = paj7620_select_register_bank(dev, PAJ7620_MEMBANK_0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	LOG_DBG("Sample rate set to %s mode", fps == PAJ7620_GAME_SPEED ? "game" : "normal");
+
+	return 0;
+}
+
+static int paj7620_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	int ret;
+	struct paj7620_data *data = dev->data;
+	const struct paj7620_config *config = dev->config;
+	uint8_t gest_data[2];
+
+	if (chan != SENSOR_CHAN_ALL
+		&& ((enum sensor_channel_paj7620)chan != SENSOR_CHAN_PAJ7620_GESTURES)) {
+		return -ENOTSUP;
+	}
+
+	/* We read from REG_INT_FLAG_1 and REG_INT_FLAG_2 even on polling mode
+	 * (without using interrupts) because that's where the gesture
+	 * detection flags are set.
+	 * NOTE: A set bit means that the corresponding gesture has been detected
+	 */
+
+	ret = i2c_burst_read_dt(&config->i2c, PAJ7620_REG_INT_FLAG_1, gest_data, sizeof(gest_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read gesture data");
+		return ret;
+	}
+
+	data->gesture_flags = sys_get_le16(gest_data);
+
+	return 0;
+}
```

Comment:
I'm concerned this attempts to force the fetch/get API to an event-driven pattern. Why not instead implement this as a streaming trigger?